plugins/vbe/fu-vbe-simple-device.c

```diff
@@ -0,0 +1,838 @@
+	last->finish_time = time(NULL);
+	last->cur_version = g_strdup(version);
+	last->status = g_strdup("completed");
+
+	return TRUE;
+}
+
+static gboolean
+fu_vbe_simple_device_write_firmware(FuDevice *device,
+				    FuFirmware *firmware,
+				    FuProgress *progress,
+				    FwupdInstallFlags flags,
+				    GError **error)
+{
+	struct _FuVbeSimpleDevice *dev = FU_VBE_SIMPLE_DEVICE(device);
+	g_autoptr(GBytes) bytes = NULL;
+	struct fit_info fit;
+	const guint8 *buf;
+	gsize size = 0;
+	gint ret;
+
+	fu_progress_set_id(progress, G_STRLOC);
+	bytes = fu_firmware_get_bytes(firmware, error);
+	if (!bytes)
+		return FALSE;
+	buf = g_bytes_get_data(bytes, &size);
+
+	g_debug("Size of FIT: %#zx\n", size);
+	ret = fit_open(&fit, buf, size);
+	if (ret) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_READ,
+			    "Failed to open FIT: %s",
+			    fit_strerror(ret));
+		return FALSE;
+	}
+
+	if (!process_fit(&fit, dev, progress, error))
+		return FALSE;
+	g_debug("write done");
+
+	/* success */
+	return TRUE;
+}
+
+static GBytes *
+fu_vbe_simple_device_upload(FuDevice *device, FuProgress *progress, GError **error)
+{
+	struct _FuVbeSimpleDevice *dev = FU_VBE_SIMPLE_DEVICE(device);
+	g_autoptr(GPtrArray) chunks = NULL;
+	gsize blksize = 0x100000;
+	gpointer buf;
+	GBytes *out;
+	off_t offset;
+	gint ret;
+
+	g_log(G_LOG_DOMAIN, G_LOG_LEVEL_INFO, "upload");
+
+	/* notify UI */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_status(progress, FWUPD_STATUS_DEVICE_READ);
+
+	ret = lseek(dev->fd, dev->area_start, SEEK_SET);
+	if (ret < 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_READ,
+			    "Cannot seek file '%s' (%d) to %#jx (%s)",
+			    dev->devname,
+			    dev->fd,
+			    (uintmax_t)dev->area_start,
+			    strerror(errno));
+		return NULL;
+	}
+
+	chunks = g_ptr_array_new_with_free_func((GDestroyNotify)g_bytes_unref);
+
+	for (offset = 0; offset < dev->area_size; offset += blksize) {
+		g_autoptr(GBytes) chunk = NULL;
+		gsize toread;
+
+		buf = g_malloc(blksize);
+		toread = blksize;
+		if ((off_t)toread + offset > dev->area_size)
+			toread = dev->area_size - offset;
+		g_log(G_LOG_DOMAIN, G_LOG_LEVEL_INFO, "read %zx", toread);
+		ret = read(dev->fd, buf, toread);
+		if (ret < 0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_READ,
+				    "Cannot read file '%s' (%s)",
+				    dev->devname,
+				    strerror(errno));
+			g_free(buf);
+			return NULL;
+		}
+		chunk = g_bytes_new_take(buf, ret);
+		g_ptr_array_add(chunks, g_steal_pointer(&chunk));
+	}
+
+	out = fu_dfu_utils_bytes_join_array(chunks);
+	g_debug("Total bytes read from device: %#zx\n", g_bytes_get_size(out));
+
+	return out;
+}
+
+static void
+fu_vbe_simple_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 100, "write");
```

Comment:
I think we're expecting 4 `fu_progress_add_step()`'s here

Reply:
I'll add a test for this.

Reply:
Done in the other PR.

Reply:
Can you explain that a bit? In this case we only do a write.

Reply:
So are you saying we need to detach from the MMC? This is not a USB device so I don't think we can do that. It is really just a case of writing to the MMC.

As for the reload, is that to make the device reload its firmware? This is the main firmware for the device (potentially for AP, EC and even perhaps some secure bits), so the only way to reload it is to reboot.

Reply:
> So are you saying we need to detach from the MMC

Reply:
Ah OK, so there always must be these four steps. I'll update it.

Reply:
Thanks for all the info. Done

I filed #4755 too.